src/python/pants/backend/jvm/tasks/jvm_compile/zinc/zinc_compile.py

```diff
@@ -388,16 +408,70 @@ def relative_to_exec_root(path):
         fp.write(arg)
         fp.write(b'\n')
 
-    if self.runjava(classpath=[self._zinc.zinc],
-                    main=Zinc.ZINC_COMPILE_MAIN,
-                    jvm_options=jvm_options,
-                    args=zinc_args,
-                    workunit_name=self.name(),
-                    workunit_labels=[WorkUnitLabel.COMPILER],
-                    dist=self._zinc.dist):
-      raise TaskError('Zinc compile failed.')
-
-  def _verify_zinc_classpath(self, classpath):
+    if self.execution_strategy == self.HERMETIC:
+      zinc_relpath = fast_relpath(self._zinc.zinc, get_buildroot())
+
+      snapshots = [
+        self._zinc.snapshot(self.context._scheduler),
+        ctx.target.sources_snapshot(self.context._scheduler),
+      ]
+
+      directory_digests = tuple(
+        entry.directory_digest for entry in dependency_classpath if entry.directory_digest
+      )
+      if len(directory_digests) != len(dependency_classpath):
+        for dep in dependency_classpath:
+          if dep.directory_digest is None:
+            logger.warning(
+              "ClasspathEntry {} didn't have a DirectoryDigest, so won't be present for hermetic "
+              "execution".format(dep)
+            )
+
+      if scala_path:
```

Comment:
Could move this into the memoized construction of `self.scala_path`, or next to it? `@memoized_method` should work in this case.